rfd/0001-kubernetes-manager.md

```diff
@@ -0,0 +1,100 @@
+state: draft
+---
+
+# RFD 1 - Kubernetes manager
+
+## What
+
+Teleport Kubernetes Manager is a [controller for Kubernetes](https://kubernetes.io/docs/concepts/architecture/controller/) serving Teleport resources defined as [Kubernetes custom resources](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+
+## Why
+
+Automated bootstrap of Teleport data is not an easy task, especially when it's running in Kubernetes. One can use `tctl` from the Teleport CLI toolset to bootstrap users/roles but integrate it into something like [Helm charts](https://helm.sh/docs/topics/charts/) quickly becomes a pain. In many scenarios, it's far more convenient to have the Teleport data defined the same way as the Kubernetes resources like Deployments and Services, in a format very similar to Teleport's own format.
+
+## Details
+
+### Implementation
+
+For the task, the [Kubebuilder](https://kubebuilder.io/) framework should be used. It has great code generation capabilities and gives a complete setup to develop & deploy the manager.
+
+### Instance definitions
+
+To write something into Teleport, the manager needs to know what server to connect to and how. So it needs Teleport Auth/Proxy/Tunnel address and credentials to be recognized. 
+
+Instance definitions look like this:
+
+```yaml
+apiVersion: instances.goteleport.com/v7
+kind: Instance
+metadata:
+  name: some-teleport-instance
+  namespace: teleport
+spec:
+  addr: some-teleport.svc.cluster.local:3080
+  secretName: some-creds
+```
+
+Credentials are stored separately from the instance definition in a secret defined in `secretName` field. An instance's secret must reside in the same namespace as an instance resource itself.
+
+Credentials data contains some private keys, and it's a common practice to store such things in a Kubernetes secret. Such separation also plays nice with Kubernetes RBAC — one would want to give access to read an instances list and introspect their status but don't give access to their secrets.
+
+Then one can create a credentials secret with something like that:
+
+```bash
+tctl auth sign --user=kubernetes-manager --ttl=8760h --overwrite --format=file -o /tmp/kubernetes-manager-identity
+kubectl -n teleport create secret generic some-creds --from-file=identity=/tmp/kubernetes-manager-identity
+rm -f /tmp/kubernetes-manager-identity
+```
+
+Secret in Kubernetes is a key-value object, and the only supported key, for now, is `"identity"` that stores Teleport connection credentials.
+
+In the future, the process of identity derivation might be simplified by adding to `tctl auth sign` support of writing an identity directly to Kubernetes instead of a temporary file.
+
+### Resource definitions
+
+Custom resource definitions of Teleport resource look like this:
+
+```yaml
+apiVersion: resources.goteleport.com/v2
+kind: Role
+metadata:
+  name: folks
+  namespace: teleport
+instanceName: some-instance
```

Comment:
I don't know if you can actually add new top-level fields to a CRD in k8s. 

You can probably achieve the same using [owner references](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/)

Reply:
> I don't know if you can actually add new top-level fields to a CRD in k8s.

I think top-level fields are not disallowed and such setup works fine in my draft implementation. `instanceName` was put on a top-level intentionally. Otherwise, definitions will look like this:

```yaml
metadata:
  name: folks
  namespace: teleport
spec:
  instanceName: some-instance
  spec:
    allow: ...
```

So it creates an unnecessary nesting and also it's confusing when you include `spec` inside `spec`. Okay `spec.spec` could be a `spec.role` but I wanted to achieve such a UX that whenever the user sees `spec` then it means the same thing as the `spec` in Teleport. This mitigates the confusion a lot.

> You can probably achieve the same using owner references

i thought about using owner references but are they supposed to be edited by the user? owner reference needs a resource UUID — where the user supposed to get it? and looks more verbose in YAML.

I always thought that owner references are for the stuff one creates programmatically.

Reply:
Yeah ownerReference is mainly for controllers (daemon/replica sets, etc.) to inject their reference into their managed pods.

Why do resources need instanceName reference at all? K8s operator can just query the "instance" resource and then create a teleport resource with it?

Reply:
Is this instance name field still needed?